clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkClient.java

```diff
@@ -272,6 +272,7 @@ private boolean trySend(long now) {
 
     private void clientPoll(long timeout, long now) {
         client.poll(timeout, now);
+        now = time.milliseconds();
```

Comment:
Also, did you keep following these calls up the call chain? I think there are some cases where the current time isn't properly updated. For example, if you look for usages of `clientPoll`, `pollUnsentRequests` uses it in a loop, where the local `now` value in that method is never updated. Maybe we just want to handle those in a separate JIRA since KAFKA-2615 was specifically about `poll()`, but I think there's still potentially a bunch of fallout from the realization that we're not updating the current time properly.

I'm hesitant to suggest this because I think it'll end up requiring a huge patch (but hopefully a straightforward one), but the way I've handled this in the past is to not pass around `long now`, but instead always pass around an object representing time. Unlike the `Time` class we have now, it explicitly exposes both a cached value and a mechanism to force an update. Something like:

```
class Time {
  private long now;

  /** Get the most recent cached value of the current time. */
  long now() {
    return now;
  }

  /** Explicit request to update the current cached time value, to be invoked after operations that could block for some time. */
  long updateNow() {
    now = System.currentTimeMillis()
    return now;
  }
}
```

This way, the `poll()` method can call `updateNow()` and everything that runs after it will see the updated value. In most cases, you can avoid assigning to a local variable at all, just using `Time.now()` to get the current value. That makes most code not need to think about exactly when the value is updated and means that if we keep following the call chain up three or four levels, we don't have to keep worrying about where exactly poll is called such that we have to update each method's local `now` value.

This approach also allows you to avoid passing `now` to all of these methods -- if you pass in the time object at construction, they can always store it in a field and use that instead.